### ISSUE TITLE
Build scripts should automatically create configuration-specific libs build log files

### ISSUE THREAD

**Issue by BruceForstall** [dotnet/runtime#43649]
Currently, the libraries build does not create any log files by default (e.g., when building locally, not in the CI system).

Apparently, when the CI system passes the "-ci" argument to the build script, that causes binlog files to be generated to the `artifacts\log\buildConfig` directory, due to it passing the "/bl" flag to msbuild.

Also, apparently, if you pass this manually, subsequent builds will use the same log file names and hence overwrite the previous log files. This is a problem if you build multiple platform/configurations, especially when building using the `build -arch x64,x86,arm,arm64` multi-arch syntax.

This should be improved, by:
1. Always generate log files for all parts of the build (the CoreCLR build already does this; I'm mainly talking about the libs build here)
2. Log files should be named by the configuration (OS, processor, build flavor) so as to not be overwritten by other builds.

@dotnet/runtime-infrastructure

**Comment by safern** [comment]
> Always generate log files for all parts of the build (the CoreCLR build already does this; I'm mainly talking about the libs build here)

I don't agree with this statement... I think binlogs should be intentionally created via a parameter, like `ci` or `bl`. We shouldn't be bloating people's file system with a log per library, or huge logs by the builds.

> Log files should be named by the configuration (OS, processor, build flavor) so as to not be overwritten by other builds.

This is arcade default behavior. We've tried to propose an update in the behavior where we can override the logs path and names, but that has always being rejected. 

What we could do is parse the `-bl` parameter in our build scripts and then pass down an msbuild parameter `/bl:<log-path` that includes the configuration and arch.

**Comment by BruceForstall** [comment]
> I think binlogs should be intentionally created via a parameter, like ci or bl. We shouldn't be bloating people's file system with a log per library, or huge logs by the builds.

How large are the log files? Is the size significant? (e.g., compared to the source code + built artifacts) The coreclr build always creates all logs. I don't recall ever hearing a concern about that. I'm ok with an opt-out switch, but IMO the default should be log files enabled. Otherwise, if you have a build failure and it scrolls off our screen/buffer, it's lost. That doesn't make sense to me.

**Comment by safern** [comment]
If you get a build failure in libraries it shouldn't scroll off the screen/buffer as we fixed the issue where it didn't fail in the first error of the build. Am I seeing something wrong here?

Maybe this is just a bug of the multi-arch build, and the multi-arch build should just fix this by passing `/bl:<path-to-binlog>` ?

**Comment by BruceForstall** [comment]
My scenario is to build as much as possible overnight, so I don't want to stop on the single first failure; I want to keep building other things. Yes, I use the multi-arch build option to build all architectures. In the failure cases I've observed, simply rebuilding has worked fine, but because there are no log files, I don't know what the original failure was.

> Maybe this is just a bug of the multi-arch build, and the multi-arch build should just fix this by passing `/bl:<path-to-binlog>` ?

That would be a start, but I still believe that should always be the default even for non-multi-arch builds.

**Comment by safern** [comment]
For non-multi-arch builds, it should stop on first failure, so what would you gain about having logs if you can look at it at the console? That is what I'm not getting. I mean, I don't mind generating a binlog for every build... but how useful would that be and we would just be slowing down people's build because binlogs impacts the build perf as well.

cc: @ViktorHofer and @ericstj for their opinion also.

**Comment by jkoritzinsky** [comment]
Maybe we can capture the output from the build and re-print out the error logs at the end (grouped by configuration/arch) when in a multi-config multi-arch build?

**Comment by BruceForstall** [comment]
> For non-multi-arch builds, it should stop on first failure, so what would you gain about having logs if you can look at it at the console?

In my experience, finding the source of a problem is often not possible just by seeing the failure messages at the end of a build, when the failure occurs. You often need to look for things that happened much earlier. And that is assuming that the build does indeed 100% stop in all failure modes at the exact point of failure. Also, console output is hopefully much less detailed than what would be learned by a more detailed build/failure log. E.g., the coreclr console output is terse, but the generated .log files are much more detailed/verbose (and the .wrn/.err files are there for future reference).

> we would just be slowing down people's build because binlogs impacts the build perf as well

I assume that's true, but it would be interesting to know how large the impact is.

**Comment by ViktorHofer** [comment]
FWIW https://github.com/dotnet/runtime/issues/43053 is related as we discussed lowering the libraries native build console's verbosity. If we would still produce verbose logs in the background I think there isn't much harm in lowering the native build's verbosity.

>  assume that's true, but it would be interesting to know how large the impact is.

+1. I doubt that build times would be noticeably (> 10s) longer. But someone should definitely do a measurement.

**Comment by safern** [comment]
I think for the multi-arch build makes total sense to enable binlogs by default... for regular builds I like that it is optional, even when there is a failure it is pretty simple to repro with one command and get a minimal repro and capture a lightweight binlog... but that is just my 2 cents.
